### PR TITLE
Add support_ignore_nulls and support_ordering in Aggregate expression

### PR DESCRIPTION
**Pull request by huaxingao**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/9924.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/group_by.slt [hunk 1]
```diff
@@ -2129,41 +2129,11 @@ query III
...
-Projection: annotated_data_infinite2.a, annotated_data_infinite2.d, SUM(annotated_data_infinite2.c) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST] AS summation1
---Aggregate: groupBy=[[annotated_data_infinite2.d, annotated_data_infinite2.a]], aggr=[[SUM(CAST(annotated_data_infinite2.c AS Int64)) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST]]]
-----TableScan: annotated_data_infinite2 projection=[a, c, d]
-physical_plan
-ProjectionExec: expr=[a@1 as a, d@0 as d, SUM(annotated_data_infinite2.c) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST]@2 as summation1]
---AggregateExec: mode=Single, gby=[d@2 as d, a@0 as a], aggr=[SUM(annotated_data_infinite2.c)], ordering_mode=PartiallySorted([1])
-----StreamingTableExec: partition_sizes=1, projection=[a, c, d], infinite_source=true, output_ordering=[a@0 ASC NULLS LAST]
-
-query III
+statement error DataFusion error: This feature is not implemented: ORDER BY is not implemented for SUM
```
Comment (huaxingao) [inline datafusion/sqllogictest/test_files/group_by.slt:2132]: I am not sure if I understand the requirement of https://github.com/apache/arrow-datafusion/issues/9924 correctly:

In my PR, I have `support_ordering` to true for `first`, `last`, `nth_value` and `array_agg_ordered`. For all the other aggregate functions, `support_ordering` is false and `ORDER BY` returns `not implemented` error.

Is this what we want? This seems to be a breaking change for me.

Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/group_by.slt:2132]: I think we should rewrite the test that excluded ordering for SUM in this case 🤔

File: datafusion/physical-expr/src/aggregate/build_in.rs [hunk 2]
```diff
@@ -417,7 +417,18 @@ pub fn create_aggregate_expr(
         (AggregateFunction::StringAgg, true) => {
             return not_impl_err!("STRING_AGG(DISTINCT) aggregations are not available");
         }
-    })
+    };
+
+    let agg_name = aggregate_expr.name();
+    if ignore_nulls && !aggregate_expr.support_ignore_nulls() {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/aggregate/build_in.rs:423]: I would prefer these checks at the beginning of the function, so we can avoid unnecessary compute for invalid cases

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 3]
```diff
@@ -3428,3 +3428,16 @@ SELECT LAST_VALUE(column1 ORDER BY column2 DESC) IGNORE NULLS FROM t;
 
 statement ok
 DROP TABLE t;
+
+# Test for IGNORE NULLS / ORDER BY not implemented
+statement error DataFusion error: This feature is not implemented: IGNORE NULLS is not implemented for COUNT
+SELECT COUNT(*) IGNORE NULLS FROM (values (1), (null), (2));
```
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: If its doable in parser we probably may want to revert those checks/tests from DF

[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: Filed https://github.com/sqlparser-rs/sqlparser-rs/issues/1206

[the remark this answers is not in the document]
Comment (huaxingao) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: I agree it is probably better to check on parser. Spark throws Exception on parser if `IGNORE NULLS` is not supported.

[the remark this answers is not in the document]
Comment (andygrove) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: I assume that we should still have the checks in the physical plan, for use cases such as Comet where we are not using the DataFusion SQL parsing?

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: I think we need to check in the physical plan layer as well for the reason @andygrove  says.

Another approach could be an analyzer rule (following the model of how certain subqueries are handled) like this:
 
https://github.com/apache/arrow-datafusion/blob/582050728914650c6d4340ca803a0e9af087d8ec/datafusion/optimizer/src/analyzer/subquery.rs#L32-L39

File: datafusion/sqllogictest/test_files/group_by.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/group_by.slt:2132]: @mustafasrepo  / @ozankabak  / @metesynnada  can you help us understand if the ordering for `SUM(c ORDER BY a DESC)` has meaning?

DataFusion currently ignores cases

We propose making it an error to include clauses that make no sense

However, I could see an argument for permitting the user to write `ORDER BY` even if the aggregate didn't care about ordering (and simply remove the ORDER BY as a optimization) 

I can also see the rationale for failing fast and erroring as the user probably didn't meant to specify an ordering on SUM 🤔

**Review by alamb — COMMENTED** [review]
Thanks @huaxingao  -- I agree this does what the ticket requests; However given what you have found about `SUM(ORDER BY ...)` it may make sense to permit that case 🤔

File: datafusion/sqllogictest/test_files/group_by.slt (hunk 1 above)
Comment (ozankabak) [reply datafusion/sqllogictest/test_files/group_by.slt:2132]: This depends on the data type. For integral types, `ORDER BY` doesn't matter for `SUM` (but it shouldn't be an error to still specify it -- the optimizer should just remove it). For floating-point types, it actually makes a difference in the result.

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 3 above)
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/sqllogictest/test_files/aggregate.slt:3434]: Queries like that should fail by the invalid syntax and this is parsers work to stop it IMHO. And spark doesn't allow query like that, it fails on query compile stage. 

```
scala> spark.sql("SELECT COUNT(*) IGNORE NULLS over () FROM (values (1), (null), (2));").show(false)
org.apache.spark.sql.AnalysisException: Function count does not support IGNORE NULLS.; line 1 pos 7
  at org.apache.spark.sql.errors.QueryCompilationErrors$.functionWithUnsupportedSyntaxError(QueryCompilationErrors.scala:602)
```

**Comment by comphead** [comment]
> Thanks @huaxingao -- I agree this does what the ticket requests; However given what you have found about `SUM(ORDER BY ...)` it may make sense to permit that case 🤔

this query is invalid, our parser allows it but it should fail. IGNORE NULLS have a limited functions to be worked with.  all other query engine parsers does.

**Comment by alamb** [comment]
> this query is invalid, our parser allows it but it should fail. IGNORE NULLS have a limited functions to be working with. all other query engine parsers does.

`IGNORE NULLS` definitely makes sense to error if the aggregate doesn't handle it (as it would result in incorrect answers)

`ORDER BY` seems more gray to me -- as in the answers won't be incorrect but it would potentially be doing more work than necessary

File: datafusion/sqllogictest/test_files/group_by.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/group_by.slt:2132]: I hadn't thought about the implications for `SUM(ORDER BY ...)` for floats 🤔  It appears that this is consistent with what postgres does:


```sql
postgres=# create table foo (x float);
CREATE TABLE
postgres=# insert into foo values (1.0);
INSERT 0 1
postgres=# insert into foo values (2.0);
INSERT 0 1
postgres=# insert into foo values (-1.0);
INSERT 0 1
postgres=# select sum(x ORDER BY x) from foo;
 sum
-----
   2
(1 row)

postgres=# select sum(x IGNORE NULLS) from foo;
ERROR:  syntax error at or near "IGNORE"
LINE 1: select sum(x IGNORE NULLS) from foo;
                     ^
postgres=#
```

I also verified that postgres actually does sort the input:
```sql
postgres=# explain select sum(x ORDER BY x) from foo;
                            QUERY PLAN
-------------------------------------------------------------------
 Aggregate  (cost=169.81..169.82 rows=1 width=8)
   ->  Sort  (cost=158.51..164.16 rows=2260 width=8)
         Sort Key: x
         ->  Seq Scan on foo  (cost=0.00..32.60 rows=2260 width=8)
(4 rows)

postgres=#
```

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Comment by huaxingao** [comment]
@alamb 

I am not so sure how to proceed with this PR. Do we still need this PR?
For `IGNORE NULLS`, I think we have consensus to error in the parser if aggregates don't support `IGNORE NULLS`.

For `ORDER BY`, there are two options:
1. Keep the current behavior, that is, allow `SUM(ORDER BY ...)`. The rationale is:

- The DataFusion optimizer would remove the `ORDER BY` if the order doesn't matter.
- For float type, the order matters, so we need the `ORDER BY`
- Postgres allows the `SUM(ORDER BY ...)`.

2. Error in the parser if ordering doesn't matter (That's what Spark does).

So, we will either change the parser or leave the current implementation as is. It seems to me we don't need this PR anymore?

**Comment by ozankabak** [comment]
> For `ORDER BY`, there are two options:
> 
> 1. Keep the current behavior, that is, allow `SUM(ORDER BY ...)`. The rationale is:
> 
> * The DataFusion optimizer would remove the `ORDER BY` if the order doesn't matter.
> * For float type, the order matters, so we need the `ORDER BY`
> * Postgres allows the `SUM(ORDER BY ...)`.

Yes, let's follow this approach.

**Comment by alamb** [comment]
> For IGNORE NULLS, I think we have consensus to error in the parser if aggregates don't support IGNORE NULLS.

Sounds good to me -- we can also move the error later if we want to enforce it for the `expr_fn` API as well

Thanks @huaxingao  and @ozankabak

**Pull request closed without merging**